CharlieBackend.AdminPanel/HtmlHelpers/CalendarHelper.cs

```diff
@@ -0,0 +1,198 @@
+
+            if (calendar.ScheduledEventFilter.FinishDate.HasValue)
+            {
+                finish = calendar.ScheduledEventFilter.FinishDate.Value;
+            }
+            else
+            {
+                DateTime latestFinish = eventOccurencesFiltered.Max(x => x.EventFinish);
+                DateTime latestStart = eventOccurencesFiltered.Max(x => x.EventStart);
+
+                finish = latestFinish > latestStart ? latestFinish : latestStart;
+            }
+
+            if (calendar.ScheduledEventFilter.StartDate.HasValue)
+            {
+                start = calendar.ScheduledEventFilter.StartDate.Value;
+            }
+            else
+            {
+                DateTime earliestFinish = eventOccurencesFiltered.Min(x => x.EventFinish);
+                DateTime earliestStart = eventOccurencesFiltered.Min(x => x.EventStart);
+
+                start = earliestFinish > earliestStart ? earliestStart : earliestFinish;
+            }
+
+            double daysCount = (finish - start).TotalDays;
+            double rowCount = daysCount / 7;
+
+            if (rowCount - (int)rowCount != 0)
+                rowCount = (int)rowCount + 1;
+
+            if (finish.DayOfWeek != DayOfWeek.Saturday)
+                daysCount += 7 - (int)finish.DayOfWeek;
+
+            List<TagBuilder> dayContainers = new List<TagBuilder>();
+
+            for (int day = 0; day < daysCount; day++)
+            {
+                dayContainers.Add(GetDayContainerHtml(start.AddDays(day), calendar.ScheduledEvents));
+            }
+
+            int startDay = (int)start.DayOfWeek;
+
+            if (start.DayOfWeek != DayOfWeek.Sunday)
+                for (int i = 1; i <= startDay; i++)
+                {
+                    dayContainers = dayContainers.Prepend(GetDateContainerHtml(start.AddDays(-i))).ToList();
+                }
+
+            List<TagBuilder> rowContainers = new List<TagBuilder>();
+
+            for (int i = 0; i < rowCount; i++)
+            {
+                rowContainers.Add(GetRowHtml(i, dayContainers));
+            }
+
+            string result = string.Empty;
+
+            foreach (var row in rowContainers)
+            {
+                using (var writer = new System.IO.StringWriter())
+                {
+                    row.WriteTo(writer, HtmlEncoder.Default);
+
+                    result += writer.ToString();
+                }
+            }
+
+            return new HtmlString(result);
+        }
+
+        public static TagBuilder GetRowHtml(int row, List<TagBuilder> days)
+        {
+            int len = row * 7 + 7;
+            TagBuilder rowBlock = new TagBuilder("div");
+            rowBlock.AddCssClass("row");
+
+            try
+            {
+                for (int i = row * 7; i < len; i++)
+                {
+                    rowBlock.InnerHtml.AppendHtml(days[i]);
+                }
+            }
+            catch
```

Comment:
Probably we should catch here something

Reply:
Done 040e96f9bd5354c667b4a1762972d7f11d64be5f